### PR TITLE
[WIP?] Fix #1 and #2

### LINKED ISSUE #1: Run with commandline
It should be possible to run this through something like `vendor/bin/php-opttimize src/`.

Preferably with an option to where to write new files to (We don't want to overwrite files).

### PR DESCRIPTION
**Pull request by Lucky-Loek**
While fixing #1, I found out that constants in the form of

`self::CONSTANT = VALUE` aren't detected.

I'll add a test case later, so you can see what's happening. Commands are testless, since I have no experience in unit testing commands. Manual testing is working.

There's also some TODOs scattered in the code. Glad to hear your feedback.

### CODE REVIEW

**Comment by BackEndTea** [comment]
Thanks for the help, i'll give it a better look later on, for now, could you run `make cs-fix` to automatically fix style errors, and run `make analyze --keep-going` to see what errors php-stan reports, and see if you can fix those.

**Comment by Lucky-Loek** [comment]
Fixed what I could. The unit test is deliberately failing so you can see what's happening. Rest is working and I've fixed a couple of bugs that prevented the .phar from running reliably. It does now!

image

File: src/Compiler.php [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+class Compiler
+{
+    /**
+     * @var string
+     */
+    private $versionDate;
+
+    public function compile(string $filename = 'php-optimize.phar'): void
+    {
+        // TODO figure out what php compatibility should be, now it's 7+ due to typehint
```
Comment (BackEndTea) [inline src/Compiler.php]: PHP compatibility is 7.1+

File: src/Compiler.php [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+
+use Seld\PharUtils\Timestamps;
+use Symfony\Component\Finder\Finder;
+
+/**
+ * Compiler class creates a PHAR file from this package. Heavily inspired by Composer's Compiler class.
+ *
+ * @author Loek van der Linde <lind0077@hz.nl>
+ */
+class Compiler
```
Comment (BackEndTea) [inline src/Compiler.php:23]: I don't have much experience with creation of phars, but what is the reason of creating our own, instead of using something like box?

File: tests/PhpParser/NodeVisitor/ConvertorVisitorTest.php [hunk 3]
```diff
@@ -51,12 +51,12 @@ public function providesConvertableCode(): \Generator
         $parser = (new ParserFactory())->create(ParserFactory::PREFER_PHP7, $lexer);
 
         yield 'It converts the use of a fully classified const' => [
-            ['Foo\Bar\Baz' => new Node\Scalar\String_('A value')],
+            ['Baz' => new Node\Scalar\String_('A value')],
```
Comment (BackEndTea) [inline tests/PhpParser/NodeVisitor/ConvertorVisitorTest.php]: Could you revert this change, and create an issue for this bug?

File: bin/compile [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+use PhpOptimizer\Compiler;
+
+try {
+    $compiler = new Compiler();
+    $compiler->compile();
+} catch (\Throwable $e) {
+    echo 'Failed to compile phar: [' . get_class($e) . '] '
+    . $e->getMessage() . ' at ' . $e->getFile() . ':' . $e->getLine() . PHP_EOL;
+    exit(1);
+}
```
Comment (BackEndTea) [inline bin/compile]: Please add a new line at the end of the file

File: bin/php-optimize [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+require __DIR__ . '/../vendor/autoload.php';
+
+use Symfony\Component\Console\Application;
+
+$application = new Application();
+
+$application->add(new PhpOptimizer\Command\ConstantToValueCommand());
+$application->add(new PhpOptimizer\Command\RunPhpOptimizerCommand());
+
+$application->run();
```
Comment (BackEndTea) [inline bin/php-optimize]: Please add a new line at the end of the file

File: src/Command/ConstantToValueCommand.php [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+use PhpParser\NodeVisitor\NameResolver;
+use PhpParser\ParserFactory;
+use PhpParser\PrettyPrinter\Standard as StandardPrettyPrinter;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Class ConstantToValue.
```
Comment (BackEndTea) [inline src/Command/ConstantToValueCommand.php]: Please remove the `Class ConstantToValue.` comment.

File: src/Command/ConstantToValueCommand.php [hunk 7]
```diff
@@ -0,0 +1,126 @@
...
+        }
+
+        $constants = $constantsVisitor->getConstants();
+
+        $convertor = new ConverterVisitor(new ConstToValueConverter($constants));
+        $traverser->addVisitor($convertor);
+
+        $prettyPrinter = new StandardPrettyPrinter();
+
+        // TODO Now we're looping twice, this should be easier to do
```
Comment (BackEndTea) [inline src/Command/ConstantToValueCommand.php:95]: For now looping twice is fine. Later we can index all the items we want to change and only loop over those again.

File: src/Compiler.php (hunk 2 above)
Comment (Lucky-Loek) [reply src/Compiler.php:23]: I have had some bad experiences with Box and how supposedly good forks handle their dependencies. Can always change this, but creating them this way is kinda like second nature nowadays ;)

File: src/Compiler.php [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /**
+     * Adds a file to the PHAR passed. Can also remove whitespace from passed source files.
+     *
+     * @param \Phar        $phar
+     * @param \SplFileInfo $file
+     * @param bool         $strip
+     */
+    private function addFile($phar, $file, $strip = true): void
```
Comment (BackEndTea) [inline src/Compiler.php]:
```php
 private function addFile(\Phar $phar, \SplFileInfo $file, bool $strip = true): void
```

File: src/Compiler.php [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /**
+     * Get file path relative to cwd.
+     *
+     * @param \SplFileInfo $file
+     *
+     * @return string
+     */
+    private function getRelativeFilePath($file)
```
Comment (BackEndTea) [inline src/Compiler.php]:
```php
private function getRelativeFilePath(\SplFileInfo $file): string
```

File: src/Compiler.php [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    }
+
+    /**
+     * Removes whitespace from PHP source, while preserving line numbers.
+     *
+     * @param string $source
+     *
+     * @return string
+     */
+    private function stripWhitespace($source)
```
Comment (BackEndTea) [inline src/Compiler.php]:
```php
private function stripWhitespace(string $source): string
```

File: src/Compiler.php [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+        return $output;
+    }
+
+    /**
+     * Adds the php-optimizer executable from the bin/ folder.
+     *
+     * @param \Phar $phar
+     */
+    private function addBinExecutable($phar): void
```
Comment (BackEndTea) [inline src/Compiler.php]:
```php
private function addBinExecutable(\Phar $phar): void
```

File: src/Compiler.php [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+        $content = \preg_replace('{^#!/usr/bin/env php\s*}', '', $content);
+        $phar->addFromString('bin/php-optimize', $content);
+    }
+
+    /**
+     * PHP forces every PHAR to end with a stub.
+     *
+     * @return string
+     */
+    private function getStub()
```
Comment (BackEndTea) [inline src/Compiler.php]:
```php
private function getStub(): string
```

File: src/Compiler.php [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+        $finder = new Finder();
+        $finder->files()
+            ->ignoreVCS(true)
+            ->name('*.php')
+            ->name('LICENSE')
+            ->exclude('Tests')
+            ->exclude('tests')
+            ->exclude('docs')
+            ->in(__DIR__ . '/../vendor/nikic/php-parser/')
```
Comment (BackEndTea) [inline src/Compiler.php:66]: Could we jut add  `../vendor`, if we run install the dependencies with not dev?

Comment (BackEndTea) [reply src/Compiler.php:66]: On that note, if possible, could you add a make target that runs these steps?

**Review by BackEndTea — COMMENTED** [review]
It looks pretty good from what i can seen, i'll check it out on a functional level in a bit.

File: src/Compiler.php (hunk 13 above)
Comment (Lucky-Loek) [reply src/Compiler.php:66]: That requires some heavy tinkering, since I've put the necessary libraries for PHAR creation (`PharUtils` and `symfony/finder`) in the --dev corner. I'll take a look at it, but for now I'm skipping it.

**Comment by BackEndTea** [comment]
Thanks for your help @loekiedepo
